clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1330,6 +1330,27 @@ event_pre_syscall(void *drcontext, int sysnum)
         return true;
     if (BUF_PTR(data->seg_base) == NULL)
         return true; /* This thread was filtered out. */
+
+    // If we just switched modes in this thread in a block ending in a syscall,
+    // we can end up here without having traced the syscall instr, which we
+    // want to avoid.  We look for an empty (non-initial, so we do not add
+    // init_header_size) buffer with zero instrs in the current window (b/c it
+    // could be empty just due to the syscall instr filling up the prior buffer).
+    // (The converse can happen, with a tracing window ending in a syscall but
+    // the mode having changed, causing us to exit up above and not emit a marker:
```

Comment:
> the mode having changed, causing us to exit up above

You mean there's a process_and_output_buffer call before the syscall in the same block?

As far as the process_and_output_buffer in event_pre_syscall is concerned, it is after we output the marker.

Reply:
Syscalls happen between blocks, never inside, so the buffer would be output before the syscall and the mode change would be during the buffer output (that's where it counts instructions while tracing).